From 82864cfc9e157807061741ed45a8097fd6aea69b Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 8 Nov 2024 10:20:03 +0800
Subject: [PATCH 129/206] mtk: hostapd: Fix hostapd crash if setup a iface with
 a link bss failed

The crash occurs while some link bsses is traversing all the links by using
for_each_mld_link(), and hostapd access to the link bss which is already
been freed.

If hostapd setup a link bss failed, the link should be removed from
its hostapd_mld. However, the function hostapd_bss_link_deinit
doesn't remove the link bss correctly if it is the first bss and
hapd->drv_priv is null. Therefore we should refator the remove iface flow
as hostapd_remove_iface (used in wifi down cmd).

There are some cases that setup a bss may fail (e.g. afc query failed) or
trigger channel switch while hostapd is setting up other links.
The failed link would be add into hostapd_mld while driver_init().

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 6421d0d0a..9554f62fd 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2759,6 +2759,7 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 			for (;;) {
 				hapd = iface->bss[j];
 				hostapd_bss_deinit_no_free(hapd);
+				hostapd_bss_link_deinit(hapd);
 				hostapd_free_hapd_data(hapd);
 				if (j == 0)
 					break;
@@ -4138,6 +4139,7 @@ int hostapd_add_iface(struct hapd_interfaces *interfaces, char *buf)
 				goto fail;
 
 			if (hostapd_setup_interface(hapd_iface)) {
+				hostapd_bss_link_deinit(hapd_iface->bss[0]);
 				hostapd_deinit_driver(
 					hapd_iface->bss[0]->driver,
 					hapd_iface->bss[0]->drv_priv,
@@ -5589,6 +5591,9 @@ int hostapd_mld_remove_link(struct hostapd_data *hapd)
 	if (!mld)
 		return -1;
 
+	if (!(mld->active_links & BIT(hapd->mld_link_id)))
+		return 0;
+
 	dl_list_del(&hapd->link);
 	mld->num_links--;
 	mld->active_links &= ~BIT(hapd->mld_link_id);
-- 
2.45.2

